core/src/main/scala/org/bykn/bosatsu/DefRecursionCheck.scala

```diff
@@ -0,0 +1,284 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+
+import cats.implicits._
+
+/**
+ * Recursion in bosatsu is only allowed on a substructural match
+ * of one of the parameters to the def. This strict rule, along
+ * with strictly finite data, ensures that all recursion terminates
+ *
+ * The rules are as follows:
+ * 0. recursive defs may not be shadowed. This makes checking for legal recursion easier
+ * 1. until we reach a recur match, we cannot call a recursive def
```

Comment:
What is the expected behaviour if you have a variable name in an outer scope that is shadowed by the recursive def, and it is referenced before the recur match? My expectation is these checks will say this rule is broken by that reference.

Reply:
Can you write an example, and I will add a test. You mean:
```
foo = 1
recursive def foo(lst):
  fn = foo
  recur lst:
    []: 0
    [_, *tail]: foo(tail).add(1)
```
something like that?

I think this won't work since that will look like accessing foo, which is in scope, but not accessible.

Reply:
added a test, let me know if that captured your idea.

Reply:
Yeah, this is exactly what I was curious about.

The other question that occurs to me is what happens when the `recur` statement isn't the return statement. I haven't had a chance to try for myself yet.

Reply:
That’s a good example to add. It should work, although it won’t be tail recursion.